compiler/src/iree/compiler/Codegen/ExternalInterfaces/Utils.h

```diff
@@ -16,6 +18,34 @@ namespace mlir::iree_compiler::IREE {
 
 static const char kEncodingInfoAttrName[] = "encoding_info";
 
+template <typename DeviceEncodingLayoutResolverAttrInterface,
+          typename EncodingLayoutAttr>
+struct WrappedExternalModel
+    : public Codegen::LayoutAttrInterface::ExternalModel<
+          DeviceEncodingLayoutResolverAttrInterface, EncodingLayoutAttr> {
+public:
+  DictionaryAttr getConfiguration(Attribute attr) const {
+    return cast<EncodingLayoutAttr>(attr).getConfiguration();
+  }
+
+  Codegen::MaterializeEncodingInfo
+  getEncodingInfo(Attribute attr, RankedTensorType type) const {
+    const DeviceEncodingLayoutResolverAttrInterface *impl =
+        static_cast<const DeviceEncodingLayoutResolverAttrInterface *>(this);
+    // If the layout is already resolved, use it directly.
+    if (auto config = impl->getConfiguration(attr)) {
+      if (auto namedAttr = config.getNamed(kEncodingInfoAttrName)) {
+        std::optional<Codegen::MaterializeEncodingInfo> info =
+            Codegen::deserializeEncodingInfo(
+                cast<DictionaryAttr>(namedAttr->getValue()));
+        assert(info && "encoding_info is invalid");
+        return info.value();
+      }
+    }
+    return impl->getEncodingInfoImpl(attr, type);
+  }
+};
```

Comment:
I think we should add `getEncodingInfoImpl` virtual method to the base class, and ask other classes to implement the virtual function.